### PR TITLE
fix: hard limit on field size while parsing line protocol

### LINKED ISSUE #21841: Enforce 1MB field limit it influxdb 1.x
__Steps to reproduce:__

Writing very large fields to influxdb can cause issues, especially with the compactor which expects field values to be a reasonable size - full compaction can fail for a shard, for example.

To reproduce, write a single series with many 1GB fields, things will get bad quickly.

__Expected behavior:__

Writes with field values that are too big to be supported should be rejected with an error. A good practical limit is 1MB. Note that the documented limit is 64 KB here: https://docs.influxdata.com/influxdb/v1.7/write_protocols/line_protocol_reference/#data-types but the TSM engine is being regularly used in practice to store larger strings as field values.

__Actual behavior:__

Currently big writes are accepted but then cause failures during internal engine operations (especially compactions).

### PR DESCRIPTION
**Pull request by lesam**
Per https://docs.influxdata.com/enterprise_influxdb/v1.9/write_protocols/line_protocol_reference/
we only support 64KB, but 1MB is a more realistic practical limit. Before this commit there was
no enforcement of field value size.

Closes #21841 

Describe your proposed changes here.

- [x] [CHANGELOG.md](https://github.com/influxdata/influxdb/blob/master/CHANGELOG.md) updated with a link to the PR (not the Issue)
- [x] [Well-formatted commit messages](https://www.conventionalcommits.org/en/v1.0.0-beta.3/)
- [x] Rebased/mergeable
- [x] Tests pass

### CODE REVIEW

**Comment by lesam** [comment]
We don't catch everywhere that can build a point, but this covers the `write` ingress which is the most important one.

**Review by gwossum — DISMISSED** [review]
Implementation and test cases looks good, have test cases for maximum possible field size and 1 greater.

It'd be cool if the test checked the error message in the HTTP body since there's multiple possible errors, but none of the other tests in server_test.go seem to check error messages. I won't hold that against this test.

**Comment by lesam** [comment]
Closing as we want to make this configurable

File: tsdb/engine.go [hunk 1]
```diff
@@ -186,9 +186,8 @@ type EngineOptions struct {
 	// nil will allow all combinations to pass.
 	ShardFilter func(database, rp string, id uint64) bool
 
-	Config         Config
-	SeriesIDSets   SeriesIDSets
-	FieldValidator FieldValidator
```
Comment (gwossum) [inline tsdb/engine.go:191]: Eliminating Engine.FieldValidator seems reasonable, since I can't find any use other than the default validator. If we get rid of Engine.FieldValidator, should we also get rid of EngineOptions.FieldValidator? There might be confusion in the future if someone thinks they can set that to change the field validation behavior.

File: tsdb/field_validator.go [hunk 2]
```diff
@@ -8,18 +8,31 @@ import (
...
+//  - the point has fields that are too long
+func ValidateFields(mf *MeasurementFields, point models.Point, skipSizeValidation bool) error {
+	pointSize := point.StringSize()
 	iter := point.FieldIterator()
 	for iter.Next() {
+		if !skipSizeValidation {
+			// Check for size of field too large. Note it is much cheaper to check the whole point size
+			// than checking the StringValue size (StringValue potentially takes an allocation if it must
+			// unescape the string, and must at least parse the string)
+			if pointSize > MaxFieldValueLength && iter.Type() == models.String {
```
Comment (gwossum) [inline tsdb/field_validator.go:24]: A possible optimization here is to compare `pointSize` to `MaxFieldValueLength` before the loop, and set `skipSizeValidation = true` before starting the loop if `pointSize <= MaxFieldValueLength`. It's possible the compiler is already performing this loop invariant hoisting, but we can make sure it is. Even if the compiler is already doing the work, this might improve readability. It would look something like:

```
    pointSize := point.StringSize()
    if pointSize <= MaxFieldValueLength {
        skipSizeValidation = true
    }
    iter := point.FieldIterator()
    for iter.Next() {
        if !skipSizeValidation {
            if iter.Type() == models.String {
```

File: tsdb/engine.go (hunk 1 above)
Comment (lesam) [reply tsdb/engine.go:191]: This is in `EngineOptions` - I don't see any `FieldValidator` remaining in the code?

File: tsdb/field_validator.go (hunk 2 above)
Comment (lesam) [reply tsdb/field_validator.go:24]: I think the code as written is simpler, and I don't want to make it less simple for a single comparison - I doubt that would show up in any profiling.

It is debatable whether I should even be doing the pointSize optimization, but I think it is a clear enough win to not re-allocate a ton of strings during StringValue that it is ok to add the check.

File: tsdb/engine.go (hunk 1 above)
Comment (gwossum) [reply tsdb/engine.go:191]: I'm not sure why I thought there was an Engine.FieldValidator. I _thought_ there was a Shard.FieldValidator, typed Engine.FieldValidator, but it was really setting the value in EngineOptions.FieldValidator. My mistake, carry on!

**Review by gwossum — APPROVED** [review]
Looks good to me. Test cases for both sides of boundary condition.
